compiler/next/test/frontend/testParseAggregate.cpp

```diff
@@ -0,0 +1,402 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/frontend/Parser.h"
+#include "chpl/queries/Context.h"
+#include "chpl/uast/Class.h"
+#include "chpl/uast/Comment.h"
+#include "chpl/uast/Function.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Record.h"
+#include "chpl/uast/Union.h"
+#include "chpl/uast/Variable.h"
+
+// always check assertions in this test
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <cassert>
+
+using namespace chpl;
+using namespace uast;
+using namespace frontend;
+
+static Builder::Result parseAggregate(Parser* parser,
+                                      const AggregateDecl*& agg,
+                                      const char* testname,
+                                      const char* prog) {
+  auto parseResult = parser->parseString(testname, prog);
+  assert(parseResult.errors.size() == 0);
+  assert(parseResult.topLevelExpressions.size() == 1);
+  assert(parseResult.topLevelExpressions[0]->isModule());
+  auto module = parseResult.topLevelExpressions[0]->toModule();
+
+  // find the first AggregateDecl statement
+  const AggregateDecl* decl = nullptr;
+  for (auto stmt : module->stmts()) {
+    if (stmt->isAggregateDecl()) {
+      decl = stmt->toAggregateDecl();
+      break;
+    }
+  }
+  assert(decl);
+  agg = decl;
+  return parseResult;
+}
+
+static void checkThisFormal(const Function* fn,
+                            const char* receiverType,
+                            Formal::Intent expectIntent) {
+  assert(fn->thisFormal() != nullptr);
+  assert(fn->thisFormal() == fn->formal(0));
+  auto formal = fn->thisFormal();
+  assert(formal->typeExpression()->isIdentifier());
+  assert(formal->typeExpression()->toIdentifier()->name() == receiverType);
+  assert(formal->initExpression() == nullptr);
+  assert(formal->name() == "this");
+  assert(formal->intent() == expectIntent);
+}
+
+static void test0(Parser* parser) {
+  const AggregateDecl* agg = nullptr;
+  auto res = parseAggregate(parser, agg, "test0.chpl",
+                            "class C { }");
```

Comment:
Might be good to have a test case with multiple classes/records/unions in a single file (or a combination)